### DISCUSSION TITLE
jlink's Missing Link: API Signature Validation - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#310] [Announcements]
# jlink's Missing Link: API Signature Validation - Gunnar Morling

Discussions around Java’s jlink tool typically center around savings in terms of (disk) space.
Instead of shipping an entire JDK,
a custom runtime image created with jlink contains only those JDK modules which an application actually requires,
resulting in smaller distributables and container images.

But the contribution of jlink — as a part of the Java module system at large — to the development of Java application’s is bigger than that:
with the notion of link time it defines an optional complement to the well known phases compile time and application run-time:

Link time is an opportunity to do whole-world optimizations that are otherwise difficult at compile time or costly at run-time. An example would be to optimize a computation when all its inputs become constant (i.e., not unknown). A follow-up optimization would be to remove code that is no longer reachable.

https://www.morling.dev/blog/jlinks-missing-link-api-signature-validation/

**Comment by eamonnmcmanus** [comment]
Excellent article!

As an alternative to a Java agent, you could write your own `main` that registers your plugin before calling jlink's `main`:

```java
package com.example;

import jdk.tools.jlink.internal.Main;
import jdk.tools.jlink.internal.PluginRepository;

public final class JlinkWrapper {
  public static void main(String[] args) throws Exception {
    PluginRepository.registerPlugin(new MyJlinkPlugin());
    Main.main(args);
  }
}
```

Of course that's accessing unsupported APIs, but as you note even the plugin API is unsupported.
